Issue number 53

smurphy8 opened the pull request:
This PR got a bit out of hand because it turned out solving #53 was quite complex.  
![newradaraxisgeneratorexample](https://cloud.githubusercontent.com/assets/182668/18611965/ee1dd4b8-7d10-11e6-89f5-de2bd65446a9.png)

That picture is with the newest data. 
# New axis threshold resolution

There is a threshold that is used to combine axis under a certain amount into an "All Others" catchall.
However, over 2 different years a different set of axis may need to be included in the All others.  
Therefore a set of ground rules had to be established as to what was included and what was not.

The rule was, if a category is not part of the All Others axis for any year, it isn't allowed to be part of it at all.  

This seems to work well.
# Karma Test Suite Added:

The code that selected which categories should become part of the axis turned into something quite a bit more complex.  To help check it I really felt tests were very necessary.  

The tests are in the [_tests](https://github.com/smurphy8/openbudgetokc/tree/issue_number_53/_test) folder.

I used the [jasmine](http://jasmine.github.io/) framework to make them and [karma](https://karma-runner.github.io/1.0/index.html) to run them. 
# Help checking installation validity.

@mkchandler  and @cahiller  if you guys get a chance to pull this down and make sure the test frameworks are installable I would appreciate it.  I will add build instructions with tests to the Readme and the overview after we have it all figured out.  

It should just install with `npm install`  if it doesn't then I need to do work!
# Lack of git squash

The commits are not as squashed as they could be because I was merging along the way.  If you want me to squash over the merge points I can but wasn't sure what was best practice there.
# Sorry for the size and trouble

Sorry for messing up the last round of this axis but I think I have all the major glitches in the function of the radar figured out.  I didn't pay enough attention to the data.  

This PR got huge because I needed pieces added to fix other pieces.  I promise I don't usually do PRs like this!

smurphy8 committed:
budget-radar spec added

smurphy8 committed:
data tests added

smurphy8 committed:
Updated syntax in radar

smurphy8 committed:
simple map check done

smurphy8 committed:
both simple axis tests pass

smurphy8 committed:
cleanup composition

smurphy8 committed:
straighten out the equals signs like a weirdo

smurphy8 committed:
makeAxisFromMap function completed need tests

smurphy8 committed:
updated budget axis tests

smurphy8 committed:
getting package.json to compile more broadly

smurphy8 committed:
Merge branch 'master' into issue_number_53

smurphy8 committed:
Tests for axis creation made more complete

I have tests to ensure that the axis that are missing are added to each
set.

This will help the radar and other such things generate the correct bounds.

smurphy8 committed:
Sorted Axis to display correctly

The radar system doesn't just need like axis, it needs
identical ordering.  That is what this commit provides.

smurphy8 committed:
Merge branch 'master' into issue_number_53

mkchandler commented:
Wow @smurphy8, this is amazing stuff! You even included tests! 😄 

I will pull it down and test it out on my machine and report back.

mkchandler commented:
@smurphy8 Also, I'm not worried about squashing the commits.

mascott commented:
That graph looks awesome!

smurphy8 commented:
Okay great!  Yeah I had to do the tests, I was losing my mind :( .

On Tue, Sep 20, 2016 at 8:43 PM, Matt Chandler notifications@github.com
wrote:

> Wow @smurphy8 https://github.com/smurphy8, this is amazing stuff! You
> even included tests! 😄
> 
> I will pull it down and test it out on my machine and report back.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codeforokc/openbudgetokc/pull/63#issuecomment-248487375,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AALJjOEKAhRuSHKiIY0Rht8I2RH9ZJgFks5qsIvWgaJpZM4J_xGf
> .

## 

Scott Murphy
scottmurphy09@gmail.com

smurphy8 commented:
Thanks,

I think it is useful, I am excited about this stuff going live!

On Wed, Sep 21, 2016 at 8:43 AM, Scott Maslar notifications@github.com
wrote:

> That graph looks awesome!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codeforokc/openbudgetokc/pull/63#issuecomment-248616201,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AALJjAfZTlyMGqJ8BPk8UKnvhuSn_CpUks5qsTRogaJpZM4J_xGf
> .

## 

Scott Murphy
scottmurphy09@gmail.com

jagthedrummer commented:
Thanks, @smurphy8! Great work.

smurphy8 committed:
changed formatting to 2 sig fig

jagthedrummer commented:
Trying to get the tests to run and not having any luck.

I've already run `npm install -g karma` and `npm install -g jasmine` and then I'm trying to run `karma start`. But it looks like jasmine is not being found or something. I've verified that both `karma` and `jasmine` are installed globally and are on my `PATH`.

``` bash
$ karma start
/usr/local/lib/node_modules/karma/node_modules/di/lib/injector.js:9
      throw error('No provider for "' + name + '"!');
      ^

Error: No provider for "framework:jasmine"! (Resolving: framework:jasmine)
    at error (/usr/local/lib/node_modules/karma/node_modules/di/lib/injector.js:22:12)
    at Object.parent.get (/usr/local/lib/node_modules/karma/node_modules/di/lib/injector.js:9:13)
    at get (/usr/local/lib/node_modules/karma/node_modules/di/lib/injector.js:54:19)
    at /usr/local/lib/node_modules/karma/lib/server.js:143:20
    at Array.forEach (native)
    at Server._start (/usr/local/lib/node_modules/karma/lib/server.js:142:21)
    at invoke (/usr/local/lib/node_modules/karma/node_modules/di/lib/injector.js:75:15)
    at Server.start (/usr/local/lib/node_modules/karma/lib/server.js:103:18)
    at Object.exports.run (/usr/local/lib/node_modules/karma/lib/cli.js:280:26)
    at Object.<anonymous> (/usr/local/lib/node_modules/karma/bin/karma:3:23)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:136:18)
```

jagthedrummer commented:
After running the following commands I got things a little bit farther:

``` bash
npm install -g karma
npm install -g jasmine
npm install -g karma-chrome-launcher
npm install -g karma-jasmine
```

Now I'm seeing this:

``` bash
$ karma start
26 09 2016 19:42:49.299:WARN [karma]: No captured browser, open http://localhost:9876/
26 09 2016 19:42:49.311:INFO [karma]: Karma v1.3.0 server started at http://localhost:9876/
26 09 2016 19:42:49.312:INFO [launcher]: Launching browsers Chrome, Firefox with unlimited concurrency
26 09 2016 19:42:49.317:ERROR [launcher]: Cannot load browser "Firefox": it is not registered! Perhaps you are missing some plugin?
26 09 2016 19:42:49.317:ERROR [karma]: Found 1 load error
26 09 2016 19:42:49.318:INFO [launcher]: Starting browser Chrome
```

But it doesn't seem that any tests run.

jagthedrummer commented:
Now I ran

```
npm install karma-firefox-launcher
```

And now the tests work!

makenova commented:
@jagthedrummer with a few changes, installing the dependencies globally can be
avoided. That would have the benefit of making sure everyone who ran the
project was using the same version of the modules i.e. the ones specified in
the `package.json`. It would also make the getting started as simple as,

``` sh
git clone wubalubadubdub.git
npm install
npm test
```

All that has to be done is, add the missing dependency to the `package.json`,

``` sh
npm install -D karma-firefox-launcher
```

and change the npm test script in the `package.json` from

``` json
    "test": "echo \"Error: no test specified\" && exit 1"
```

to

``` json
    "test": "karma start"
```

If anyone is curious why that works, read [this](https://docs.npmjs.com/cli/run-script), especially the part about the PATH.

smurphy8 commented:
@makenova Thanks for explaining this!  So helpful!

mkchandler commented:
Thanks @makenova!

smurphy8 committed:
Changed package.json to include karma

Thanks to a helpful note from @makenova, found the correct way
to add karma and deps to the npm package.json file.

smurphy8 commented:
@makenova I made the change you suggested, please tell me if you see anything else that would be good to make this work better.

Anyone want to make a travis CI for our new test suite??? :smiley:

makenova commented:
@smurphy8 If there isn't a deadline, I'll have a go at it.
I've never set it up before so, It will not be fast.

smurphy8 commented:
If you run into trouble please let me know.

I don't think there is a deadline as this isn't stopping us from going into
production or anything.

On Wed, Sep 28, 2016 at 8:42 AM, Makenova notifications@github.com wrote:

> @smurphy8 https://github.com/smurphy8 If there isn't a deadline, I'll
> have a go at it.
> I've never set it up before so, It will not be fast.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codeforokc/openbudgetokc/pull/63#issuecomment-250169851,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AALJjBPVYWtCt3un5VGvgdQPEKjc13oSks5qum62gaJpZM4J_xGf
> .

## 

Scott Murphy
scottmurphy09@gmail.com

makenova commented:
@smurphy8 Looks like someone with admin access will have to enable travis on the repo.

When that is done, a `.travis.yml` file will need to be added to the root of the project.

The following is an example that will run the tests using the latest version of node.

``` tml
language: node_js
node_js:
  - "6"
```

By default, it will install dependencies with `npm install` and run tests using `npm test`.

Here is a [link](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/)  to the node documentation.

mkchandler commented:
@makenova You should have admin access to the repo now 😄

smurphy8 commented:
Awesome, hooray for CI!

On Thu, Sep 29, 2016 at 9:49 AM, Matt Chandler notifications@github.com
wrote:

> @makenova https://github.com/makenova You should have admin access to
> the repo now 😄
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codeforokc/openbudgetokc/pull/63#issuecomment-250488789,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AALJjGiZ339HFbgP5xRnEUuYRoRO6SZcks5qu8_9gaJpZM4J_xGf
> .

## 

Scott Murphy
scottmurphy09@gmail.com

makenova commented:
@mkchandler thanks!

@smurphy8 Should I wait for this pull request to be merged?

mkchandler commented:
@smurphy8 I am ok with merging this now. It has accomplished the intent, and any future modifications can be done separately. Are you ok with merging?

smurphy8 commented:
:shipit:

On Sep 29, 2016 13:40, "Matt Chandler" notifications@github.com wrote:

> @smurphey8 I am ok with merging this now. It has accomplished the intent,
> and any future modifications can be done separately. Are you ok with
> merging?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/codeforokc/openbudgetokc/pull/63#issuecomment-250554715,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AALJjDrYpsGUWRv7rDDFdyZaqnYmeJ9Kks5qvAY0gaJpZM4J_xGf
> .

jagthedrummer commented:
Thanks for the assist @makenova! This is looking good to me. Let's merge it.
